crates/lib3h/src/engine/real_engine.rs

```diff
@@ -155,20 +193,91 @@ impl<'engine> RealEngine<'engine> {
+        let gateway = if let Some(chain_id) = maybe_chainId {
+            self.space_gateway_map
+                .get_mut(&chain_id)
+                .expect("Should have the space gateway")
+        } else {
+            &mut self.network_gateway
+        };
+        gateway
+            .request(
+                Lib3hTrace,
+                GatewayRequestToChild::Dht(DhtRequestToChild::RequestThisPeer),
+                Box::new(|mut ud, response| {
+                    let response = {
+                        match response {
+                            GhostCallbackData::Timeout => panic!("timeout"),
+                            GhostCallbackData::Response(response) => match response {
+                                Err(e) => panic!("{:?}", e),
+                                Ok(response) => response,
+                            },
+                        }
+                    };
+                    if let GatewayRequestToChildResponse::Dht(
+                        DhtRequestToChildResponse::RequestThisPeer(peer_response),
+                    ) = response
+                    {
+                        ud.this_peer = peer_response;
+                    } else {
+                        panic!("bad response to bind: {:?}", response);
+                    }
+                    Ok(())
+                }),
+            )
+            .expect("sync functions should work");
+        gateway.process(&mut self.gateway_user_data).unwrap(); // FIXME unwrap
+        self.gateway_user_data.this_peer.clone()
+    }
+
+    ///
+    pub fn get_peer_list_sync(&mut self) -> Vec<PeerData> {
+        trace!("engine.get_peer_list_sync() ...");
+        self.network_gateway
+            .request(
+                Lib3hTrace,
+                GatewayRequestToChild::Dht(DhtRequestToChild::RequestPeerList),
+                Box::new(|mut ud, response| {
+                    let response = {
+                        match response {
+                            GhostCallbackData::Timeout => panic!("timeout"),
+                            GhostCallbackData::Response(response) => match response {
+                                Err(e) => panic!("{:?}", e),
+                                Ok(response) => response,
+                            },
+                        }
+                    };
+                    if let GatewayRequestToChildResponse::Dht(
+                        DhtRequestToChildResponse::RequestPeerList(peer_list_response),
+                    ) = response
+                    {
+                        ud.peer_list = peer_list_response;
+                    } else {
+                        panic!("bad response to bind: {:?}", response);
+                    }
+                    Ok(())
+                }),
+            )
+            .expect("sync functions should work");
+        self.network_gateway
+            .process(&mut self.gateway_user_data)
+            .unwrap(); // FIXME unwrap
```

Comment:
Unlike my previous comment, this one actually needs to be realtime. Ideally we'd make the functions that need this list just actually wait on the async response (i.e. the code that follows this would move inside the callback.)

If we want to side-step that to get something working faster, I'd recommend switching this to a direct function call on the concrete gateway.